# ProcessWholeGenomeAlignments

This is a preliminary projects. Eventually, I will work on to make the pipeline automatic.

 - `halLiftover.sh`   Pairwise halLiftover by running the script multiple times
 - `bedAggregation.py`   Aggregate tiny gapped alignments generated by halLiftover in progressive cactus
 - `ortholog_mapping.py`   Get the aligned orthologs in progressive cactus
 - `ortholog_relations.py`   Get the ortholog information from the pairwise aligned orthologs identified from ortholog_mapping.py
 - `find_last_orf.py`  Find genes that are aligned to unannotated regions in their outgroup species
 - `check_homologs.py`   Exclude genes that are aligned to unannotated regions, but also have homologs in other chromosome regions in outgroup species.
 - `extract_raw_hit.py`  Extract raw hits from aggregated bed files
 - `extract_alnscore_genewise.py`  Extract spliced alignment scores and get final de novo gene candidates
 - `process_spliced_aln.sh`   This combines `extract_raw_hit.py` and `extract_alnscore_genewise.py`
